### PR TITLE
MRD-1591 - ability to download last completed document

### PR DESCRIPTION
**Pull request by robinandrewdevilliers**

### CODE REVIEW

File: server/controllers/caseSummary/getCaseSection.test.ts [hunk 1]
```diff
@@ -61,6 +62,40 @@ describe('getCaseSection', () => {
     )
   })
 
+  it('caches the contact history response in redis if CRN is not excluded or restricted', async () => {
```
Comment (DarrenW-HMPPS) [inline server/controllers/caseSummary/getCaseSection.test.ts:65]: I might be missing something, but it doesn't look like this tests that the response is cached in Redis.

File: server/controllers/recommendation/countersigningTelephoneController.test.ts [hunk 2]
```diff
@@ -2,12 +2,13 @@ import { mockNext, mockReq, mockRes } from '../../middleware/testutils/mockReque
 import { getStatuses, updateRecommendation } from '../../data/makeDecisionApiClient'
 import recommendationApiResponse from '../../../api/responses/get-recommendation.json'
 import countersigningTelephoneController from './countersigningTelephoneController'
+import { STATUSES } from '../../middleware/recommendationStatusCheck'
 
 jest.mock('../../data/makeDecisionApiClient')
 
 describe('get', () => {
   it('load with no data for SPO', async () => {
-    ;(getStatuses as jest.Mock).mockResolvedValue([{ name: 'SPO_SIGNATURE_REQUESTED', active: true }])
```
Comment (DarrenW-HMPPS) [inline server/controllers/recommendation/countersigningTelephoneController.test.ts:10]: Same comment here about using production code constants in tests

File: server/controllers/recommendation/redirectController.test.ts [hunk 3]
```diff
@@ -211,8 +211,8 @@ describe('get', () => {
     expect(next).toHaveBeenCalled()
   })
 
-  it('redirect to task-list-no-recall if SPO_CONSIDERING_RECALL', async () => {
```
Comment (DarrenW-HMPPS) [inline server/controllers/recommendation/redirectController.test.ts:214]: Is this now a duplicate of the test on line 72?

File: server/controllers/recommendation/redirectController.test.ts [hunk 4]
```diff
@@ -211,8 +211,8 @@ describe('get', () => {
...
-    ;(getStatuses as jest.Mock).mockResolvedValue([{ name: 'SPO_CONSIDERING_RECALL', active: true }])
```
Comment (DarrenW-HMPPS) [inline server/controllers/recommendation/redirectController.test.ts:215]: Is it this that has prompted you to use the STATUSES enum? i.e. perhaps this used to be called SPO_CONSIDERING_RECALL but is now SPO_CONSIDER_RECALL? Or perhaps it's not that it's changed, it's just a bug in the tests.

If SPO_CONSIDERING_RECALL no longer exists, there are still a couple of references to it in the project that probably ought to be removed.

**Review by DarrenW-HMPPS — CHANGES_REQUESTED** [review]

File: server/controllers/caseSummary/caseSummaryController.test.ts [hunk 5]
```diff
@@ -415,7 +415,7 @@ describe('get', () => {
 
   it('do not show recommendation button for spo when recommendation doc and SPO_CONSIDER_RECALL state', async () => {
     ;(getCaseSummary as jest.Mock).mockReturnValueOnce(caseOverviewApiResponse)
-    ;(getStatuses as jest.Mock).mockReturnValueOnce([{ name: 'SPO_CONSIDER_RECALL', active: true }])
+    ;(getStatuses as jest.Mock).mockReturnValueOnce([{ name: STATUSES.SPO_CONSIDER_RECALL, active: true }])
```
[the remark this answers is not in the document]
Comment (robinandrewdevilliers) [reply server/controllers/caseSummary/caseSummaryController.test.ts:418]: This is a catch-up edit - the majority of the tests are using the 'enum', if the enum's not correct, then then acceptance tests are will catch it anyway.

File: server/controllers/caseSummary/getCaseSection.test.ts [hunk 6]
```diff
@@ -6,6 +6,7 @@ import {
   RecommendationResponse,
   VulnerabilitiesResponse,
 } from '../../@types/make-recall-decision-api'
+import { LastCompletedRecommendationsResponse } from '../../@types/make-recall-decision-api/models/LastCompletedRecommendationsResponse'
```
[the remark this answers is not in the document]
Comment (robinandrewdevilliers) [reply server/controllers/caseSummary/getCaseSection.test.ts:9]: fixed

File: server/controllers/caseSummary/getCaseSection.test.ts (hunk 1 above)
Comment (robinandrewdevilliers) [reply server/controllers/caseSummary/getCaseSection.test.ts:65]: wrong test name - fixed.

File: integration_tests/integration/lastCompletedList.spec.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  })
+
+  it('shows a message if no last completed', () => {
+    cy.signIn()
+    cy.task('getCase', {
+      sectionId: 'last-completed',
+      statusCode: 200,
+      response: {
+        ...getLastCompletedRecommendationsResponse,
+        recommendations: [],
```
[the remark this answers is not in the document]
Comment (robinandrewdevilliers) [reply integration_tests/integration/lastCompletedList.spec.ts:38]: null now.

**Review by DarrenW-HMPPS — APPROVED** [review]

**Pull request merged**